lib/render/SliceRenderer.cpp

```diff
@@ -0,0 +1,700 @@
+    box->GetExtents(min, max);
+    if (_cacheParams.boxMin != min)                                  return true;
+    if (_cacheParams.boxMax != max)                                  return true;
+    return false;
+}
+
+int SliceRenderer::_paintGL(bool fast) {
+    int rc = 0;
+
+    _initializeState();
+
+    if (_isDataCacheDirty()) {
+        rc = _resetDataCache();
+        if (rc<0) {
+            _resetState();
+            return rc;  // error message already set by _resetDataCache()
+        }
+    }
+    else {
+        if (_isColormapCacheDirty())
+            _resetColormapCache();
+    
+        if (_isBoxCacheDirty()) {
+            rc = _resetBoxCache();
+            if (rc<0) {
+                _resetState();
+                return rc;  // error message already set by _resetBoxCache()
+            }
+        }
+    }
+
+    _configureShader();
+    if (printOpenGLError() !=0) {
+        _resetState();
+        return -1;
+    }
+
+    glActiveTexture(GL_TEXTURE0);
+    glBindTexture(GL_TEXTURE_1D, _colorMapTextureID);
+
+    glActiveTexture(GL_TEXTURE1);
+    glBindTexture(GL_TEXTURE_2D, _dataValueTextureID);
+
+    glBindVertexArray(_VAO);
+    glDrawArrays(GL_TRIANGLES, 0, 6);
+    
+    _resetState();
+
+    if (printOpenGLError() !=0) {
+        return -1;
+    }
+    return rc;
+}
+
+void SliceRenderer::_configureShader() {
+    ShaderProgram* s = _glManager->shaderManager->GetShader("Slice");
+    s->Bind();
+    
+    // One vertex shader uniform vec4
+    s->SetUniform("MVP", _glManager->matrixManager->GetModelViewProjectionMatrix());
+
+    // Remaining fragment shader uniform floats
+    SliceParams* p = dynamic_cast<SliceParams*>(GetActiveParams());
+    assert(p);
+    float opacity = p->GetConstantOpacity();
+    s->SetUniform("constantOpacity", opacity);
+    s->SetUniform("minLUTValue", (float)_cacheParams.tf_minMax[0]);
+    s->SetUniform("maxLUTValue", (float)_cacheParams.tf_minMax[1]);
+
+    // And finally our uniform samplers
+    GLint colormapLocation;
+    colormapLocation = s->GetUniformLocation("colormap");
+    glUniform1i(colormapLocation, 0);
+
+    GLint dataValuesLocation;
+    dataValuesLocation = s->GetUniformLocation("dataValues");
+    glUniform1i(dataValuesLocation, 1);
+}
+
+void SliceRenderer::_initializeState() {
+    _glManager->matrixManager->MatrixModeModelView();
+    glEnable(GL_BLEND);
+    glEnable(GL_DEPTH_TEST);
+    glPolygonMode(GL_FRONT_AND_BACK, GL_FILL);
+    glBlendFunc(GL_SRC_ALPHA, GL_ONE_MINUS_SRC_ALPHA);
+    glDepthMask(GL_TRUE);
+}
+
+void SliceRenderer::_resetState() {
+    glActiveTexture(GL_TEXTURE1);
+    glBindTexture(GL_TEXTURE_2D, 0);
+
+    glActiveTexture(GL_TEXTURE0);
+    glBindTexture(GL_TEXTURE_1D, 0);
+   
+    glBindVertexArray(0);
+    glBindBuffer(GL_ARRAY_BUFFER, 0);
+    
+    ShaderProgram* s = _glManager->shaderManager->GetShader("Slice");
+    s->UnBind();
+    
+    glDisable(GL_BLEND);
+    glDisable(GL_DEPTH_TEST);
+    glDepthMask(GL_FALSE);
```

Comment:
While we don't have an official state (I think we should add one?) I think for the renderers the depth buffer should be enabled by default.

Reply:
We were talking about this earlier, and and at the time we decided that I should turn off everything that gets turned on during the rendering.  I do think we could sit down and talk about this and flesh out a consistent protocol though.